rpcs3/Emu/RSX/rsx_methods.cpp

```diff
@@ -819,36 +819,85 @@ namespace rsx
 		template<u32 index>
 		struct color
 		{
-			static void impl(thread* rsx, u32 _reg, u32 arg)
+			static void impl(thread* rsx, u32 /*_reg*/, u32 /*arg*/)
 			{
-				if (index >= method_registers.nv308a_size_out_x())
+				const u32 out_x_max = method_registers.nv308a_size_out_x();
+
+				if (index >= out_x_max)
 				{
 					// Skip
 					return;
 				}
 
-				u32 color = arg;
-				u32 write_len = 4;
+				// Get position of the current command arg
+				const u32 src_offset = rsx->fifo_ctrl->get_pos() - 4;
+
+				// Get real args count (starting from NV3089_COLOR)
+				const u32 count = std::min<u32>({rsx->fifo_ctrl->get_remaining_args_count() + 1,
+					static_cast<u32>(((rsx->ctrl->put & ~3ull) - src_offset) / 4), 0x700 - index, out_x_max - index});
+
+				const u32 dst_dma = method_registers.blit_engine_output_location_nv3062();
+				const u32 dst_offset = method_registers.blit_engine_output_offset_nv3062();
+				const u32 out_pitch = method_registers.blit_engine_output_pitch_nv3062();
+
+				const u32 x = method_registers.nv308a_x() + index;
+				const u32 y = method_registers.nv308a_y();
+
+				// TODO
+				//auto res = vm::passive_lock(address, address + write_len);
+
 				switch (method_registers.blit_engine_nv3062_color_format())
 				{
 				case blit_engine::transfer_destination_format::a8r8g8b8:
 				case blit_engine::transfer_destination_format::y32:
 				{
-					// Bit cast
+					// Bit cast - optimize to mem copy
+
+					const auto dst = vm::_ptr<u8>(get_address(dst_offset + (x * 4) + (out_pitch * y), dst_dma, HERE));
+					const auto src = vm::_ptr<const u8>(get_address(src_offset, CELL_GCM_LOCATION_MAIN, HERE));
+
+					const u32 data_length = count * 4;
+					const bool is_overlapping = (dst_dma & CELL_GCM_LOCATION_MAIN) && [&]() -> bool
```

Comment:
std::memmove should already do all this for you. Just use memmove and simplify this logic.

Reply:
Notice this isn't a memmove equivalent - It's executing a mem copy in u32 unit specifically as previous logic.

Reply:
But if realhw is known to handle it similarly as a memmove then ok,

Reply:
1. There is no logical difference between the two paths. You can just use memcpy for both.
2. This overlap is not a reasonable condition anyway. Nobody is scheduling inline transfers to destroy fifo that is actively being processed and expecting it to work.

The way I see it, pick one, memcpy or memmove and call it a day.

Reply:
Ok then, I only uses the destination location as a hint to choose between memcpy and memmove.